Add local git uri support for userPlugins and the abbility for userPlugins to specify their own derivation (fixed git history)

SpiderUnderUrBed opened the pull request:
This PR makes the fundamental process for adding custom user plugins better with the ability to add more git uris and more flexibility for how plugins are built.

I maybe need to do more work making sure that it actually works, and I intend to spend the next few days working on this PR to hopefully get this done despite my absence

Why I created this PR:
This PR supersedes #40 which had a messed up git history despite multiple re base attempts and cherry-picks, even when the git history was fixed it did not display correctly, I did not understand the ramifications of this so i created this PR again to be safe.

Reviewer:
@FlameFlag

SpiderUnderUrBed commented:
@DontEatOreo 
This may be a stupid question, but:
```
nixcord.homeManagerModules.nixcord
```
does not work, I am pretty sure I didnt touch the flake or did anything that would prevent this from working
https://pastebin.com/bSnftuVB
^ This is my flake, sorry its messy

DontEatOreo commented:
Could you please clean up the commit?

1. Reset it with `git reset --soft HEAD~1` and use `git add -p` to stage only the relevant changes, as it currently includes unrelated files and changes
2. Name the commit to something shorter, like `feat(lib/hm-modules): support local git URI for userPlugins`

Also as I mentioned [previously](https://github.com/KaylorBen/nixcord/pull/40#issuecomment-2974702979), please also provide some example user plugins so I can test the PR

> `nixcord.homeManagerModules.nixcord` does not work, I am pretty sure I didnt touch the flake

The flake was a bit hard to review, but I think the issue is that you're accidentally shadowing the nixcord flake on line 101

FlameFlag reviewed (changes requested):
Also again, please reset all the commits and use `git add -p` to only add relevant changes...

SpiderUnderUrBed commented:
Re-open

SpiderUnderUrBed commented:
Nope that did not work.

SpiderUnderUrBed commented:
Ah sorry, some weird git issue, ill fix it right now, as for your other questions, ill answer them soon.

SpiderUnderUrBed committed:
feat(lib/hm-modules): support local git URI for userPlugins

Eveeifyeve commented:
The git commit history looks sus.

SpiderUnderUrBed commented:
@Eveeifyeve 
Fresh after a rebase, I tried to do a force push, as I had no reason to think it would adversly affect the commit history, and for that, somehow the commit that implemented this, disappeared, and github automatically closed this PR, which is bad, but anyways, I have been working on fixing the commit history (quite close now), and mainly finishing this feature, hopefully I can push changes soon (things have been hectic with assignments)

Eveeifyeve reviewed (changes requested):
As a outsider that maintains nixpkgs and windows specific packages in nixpkgs, I can't really test this pr and the commit history needs to be fixed as said [by @FlameFlag many times](https://github.com/KaylorBen/nixcord/pull/111#issuecomment-2976185466). Until then I am adding a blocker myself, because this pr looks suspicious like malware is being implanted [(aka that one incident with linux kernal)](https://lore.kernel.org/all/CAHk-=wj4a_CvL6-=8gobwScstu-gJpX4XbX__hvcE=e9zaQ_9A@mail.gmail.com/) with the commit history and it can be fixed.

SpiderUnderUrBed commented:
I'll try fix the commit history before the end of tomorrow.

FlameFlag reviewed (commented):
This looks much better, but please also provide example plugins so I can test it on my end

Also if you haven't already, follow the advice here: https://github.com/KaylorBen/nixcord/pull/111#issuecomment-3049671764

github-actions[bot] committed:
github: update discord versions

- linux: ptb: 0.0.154 -> 0.0.155
- darwin: ptb: 0.0.184 -> 0.0.185

github-actions[bot] committed:
github: update vencord-unstable c7e799e -> 72329f9

github-actions[bot] committed:
github: update discord versions

- linux: canary: 0.0.736 -> 0.0.737
- darwin: canary: 0.0.842 -> 0.0.843

github-actions[bot] committed:
github: update vencord 1.12.9 -> 1.12.10

github-actions[bot] committed:
github: update vencord-unstable 72329f9 -> f356f64

SpiderUnderUrBed commented:
@DontEatOreo I will do that, the github bot thing is annoying me and spamming my inbox, also for user plugins you can try this:
```
vc-betterActivities
```
```
```
vc-betterActivities = "github:D3SOX/vc-betterActivities/b42afcd35d0ade108114b301859c7a077f45a8d5";
```

Clone the repo, and just specify the path to that cloned repo if you want to try the path, and you can try the other supported providers, I do have my own userplugin i am working on, and try testing, but its not really ready yet and at most you'll see it in the menu. Currently I am experiencing some weird issue:
```
error: builder for '/nix/store/n5vhv5sn46952mqy7lbv2di34p059cy4-vencord-1.12.4.drv' failed with exit code 1;
       last 25 log lines:
       >
       >     ../../nix/store/wrxi5dz8gpvrhq5aippsg8fgddsn1lww-source/index.tsx:14:72:
       >       14 │ ...opsLazy, findComponentByCodeLazy, findStoreLazy } from "@webpack";
       >          ╵                                                           ~~~~~~~~~~
       >
       >   You can mark the path "@webpack" as external to exclude it from the bundle, which will remove this error and leave the unresolved path in the bundle.
       >
       > ✘ [ERROR] Could not resolve "@webpack/common"
       >
       >     ../../nix/store/wrxi5dz8gpvrhq5aippsg8fgddsn1lww-source/index.tsx:15:107:
       >       15 │ ... UserStore, useState, useStateFromStores } from "@webpack/common";
       >          ╵                                                    ~~~~~~~~~~~~~~~~~
       >
       >   You can mark the path "@webpack/common" as external to exclude it from the bundle, which will remove this error and leave the unresolved path in the bundle.
       >
       > 7 errors
       > Build failed with 7 errors:
       > ../../nix/store/wrxi5dz8gpvrhq5aippsg8fgddsn1lww-source/index.tsx:9:60: ERROR: Could not resolve "@api/Settings"
       > ../../nix/store/wrxi5dz8gpvrhq5aippsg8fgddsn1lww-source/index.tsx:10:33: ERROR: Could not resolve "@api/Styles"
       > ../../nix/store/wrxi5dz8gpvrhq5aippsg8fgddsn1lww-source/index.tsx:11:26: ERROR: Could not resolve "@components/ErrorBoundary"
       > ../../nix/store/wrxi5dz8gpvrhq5aippsg8fgddsn1lww-source/index.tsx:12:21: ERROR: Could not resolve "@utils/constants"
       > ../../nix/store/wrxi5dz8gpvrhq5aippsg8fgddsn1lww-source/index.tsx:13:41: ERROR: Could not resolve "@utils/types"
       > ...
       > 7 errors
       >  ELIFECYCLE  Command failed with exit code 1.
       For full logs, run:
         nix log /nix/store/n5vhv5sn46952mqy7lbv2di34p059cy4-vencord-1.12.4.drv
error: 1 dependencies of derivation '/nix/store/86isk7msw1xizpbq2xvyyaf67by7g7c3-discord-0.0.99.drv' failed to build
error: 1 dependencies of derivation '/nix/store/nix9dvizxmvcdiviq7aykl6ghfhljx9n-use_system_vencord.patch.drv' failed to build
error: 1 dependencies of derivation '/nix/store/rhqjf58xyx01bcvsbl2592q7vrgc7ili-home-manager-path.drv' failed to build
error: 1 dependencies of derivation '/nix/store/cm32zr2z7gkkkfdlvdlnl9sxbq3ry9js-home-manager-generation.drv' failed to build
```
I am trying to figure it out, if you know why I am getting this issue, please tell me, otherwise ill just try to figure out why its happening. At one point I managed to substitute the library imports with the actual location to the libraries themselves, but that was sub-optimal and got some weird behavior.

SpiderUnderUrBed commented:
Its actually related to #124, maybe a fix could be addressed in this PR or another one

github-actions[bot] committed:
github: update discord versions

- linux: canary: 0.0.737 -> 0.0.739
- darwin: canary: 0.0.843 -> 0.0.845

github-actions[bot] committed:
github: update discord versions

- linux: ptb: 0.0.155 -> 0.0.156,canary: 0.0.739 -> 0.0.740
- darwin: ptb: 0.0.185 -> 0.0.186,canary: 0.0.845 -> 0.0.844

github-actions[bot] committed:
github: update dorion 6.9.1 -> 6.10.0

github-actions[bot] committed:
github: update discord versions

- darwin: canary: 0.0.844 -> 0.0.847

SpiderUnderUrBed commented:
Has the user plugins issue been fixed?
